homotopy-core/src/migration.rs

```diff
@@ -0,0 +1,324 @@
+        };
+
+        Ok(Cospan { forward, backward })
+    }
+
+    fn load_rewrite(&mut self, index: usize) -> Result<Rewrite> {
+        if let Some(rewrite) = self.rewrites.get(&index) {
+            return Ok(rewrite.clone());
+        }
+
+        // Get rewrite information
+        let dim: usize = from_value(self.stored[index][1]["n"].clone())?;
+        let cones_index: usize =
+            from_value(self.stored[index][1]["f"]["components"]["_l"].clone())?;
+        let cones_data = self.generate_vec(cones_index)?;
+
+        let rewrite: Rewrite = {
+            if dim == 0 {
+                let i: usize = from_value(cones_data[0]["_l"].clone())?;
+
+                let source: String = from_value(self.stored[i][1]["f"]["source_id"].clone())?;
+                let source: usize = source.parse().unwrap();
+                let source = self.generators[&source];
+
+                let target: String = from_value(self.stored[i][1]["f"]["target_id"].clone())?;
+                let target: usize = target.parse().unwrap();
+                let target = self.generators[&target];
+
+                Rewrite0::new(source, target).into()
+            } else {
+                let mut cones: Vec<Cone> = Vec::new();
+                for v in cones_data {
+                    let i: usize = from_value(v["_l"].clone())?;
+                    let c = self.load_cone(i)?;
+                    cones.push(c);
+                }
+                RewriteN::new(dim, cones).into()
+            }
+        };
+
+        self.rewrites.insert(index, rewrite.clone());
+        Ok(rewrite)
+    }
+
+    fn load_cone(&mut self, index: usize) -> Result<Cone> {
+        if let Some(cone) = self.cones.get(&index) {
+            return Ok(cone.clone());
+        }
+
+        // Get cone information
+        let source_index = from_value(self.stored[index][1]["f"]["source_data"]["_l"].clone())?;
+        let target_index = from_value(self.stored[index][1]["f"]["target_data"]["_l"].clone())?;
+        let slices_index = from_value(self.stored[index][1]["f"]["sublimits"]["_l"].clone())?;
+        let cone_index = from_value(self.stored[index][1]["f"]["first"].clone())?;
+
+        // Source and target
+        let source_data = self.generate_vec(source_index)?;
+        let mut source: Vec<Cospan> = Vec::new();
+        for v in source_data {
+            let i: usize = from_value(v["_l"].clone())?;
+            let c = self.load_cospan(i)?;
+            source.push(c);
+        }
+        let target = self.load_cospan(target_index)?;
+
+        // Slices
+        let sublimits = self.generate_vec(slices_index)?;
+        let mut slices: Vec<Rewrite> = Vec::new();
+        for v in sublimits {
+            let i: usize = from_value(v["_l"].clone())?;
+            let c = self.load_rewrite(i)?;
+            slices.push(c);
+        }
+
+        let cone = Cone::new(cone_index, source, target, slices);
+        self.cones.insert(index, cone.clone());
```

Comment:
Hmm what is `index` here? Cones are expected to be ordered by their index (i.e. `cone_index` above).

Reply:
That's for caching loaded cones, and `index` is the cone's array index in `stored`.

Reply:
Makes sense